SCC-2972: Add Dimensions Field to Bottom Bib Details

charmingduchess opened the pull request:
**What's this do?**
Dimensions wasn't being extracted from the bib to become a display field. Now it is. 

**Why are we doing this? (w/ JIRA link if applicable)**
[SCC-2972](https://jira.nypl.org/browse/SCC-2972)

WormJim commented on src/app/components/BibPage/BottomBibDetails.jsx:
```diff
@@ -20,6 +20,7 @@ const BottomBibDetails = ({ bib, resources }) => {
     { label: 'Publication Date', value: 'serialPublicationDates' },
     { label: 'Description', value: 'extent' },
     { label: 'Donor/Sponsor', value: 'donor' },
+    { label: 'Dimensions', value: 'dimensions'},
```
While this is perfectly fine and worthy of approval (such fanciness 🥇) I'm curious if we want to put Dimensions in this position if the array.

The list is not in alphabetical order, and it doesn't really seem to be in any particular order, so should it be placed in the 6th position, a different position, or does it not matter?

charmingduchess replied:
i asked sean and he said at the end.

WormJim replied:
ok

WormJim reviewed (approved):
OMG! This is too complex!

WormJim commented on test/unit/Bib.test.js:
```diff
@@ -10,6 +10,8 @@ import Bib, { addCheckInItems } from './../../src/server/ApiRoutes/Bib';
   /* holding could lack location, as shown in second holding */
   const mockBib = {
+    extent: ['99 bottles of beer'],
+    dimensions: ['99 x 99 cm'],
```
Can `extent` or `dimensions` have more than one index?

WormJim commented on src/server/ApiRoutes/Bib.js:
```diff
@@ -180,7 +191,7 @@ function fetchBib(bibId, cb, errorcb, reqOptions, res) {
       if (!data.annotatedMarc || !data.annotatedMarc.bib)
         data.annotatedMarc = null;
-
+      data.extent[0] = appendDimensionsToExtent(data)
```
I'm not aware, will there always be a single index to append dimensions to?

WormJim commented on test/unit/Bib.test.js:
```diff
@@ -54,6 +56,21 @@ describe('Bib', () => {
     ],
   };
 
+  describe.only('appendDimensionsToExtent', () => {
```
I think you forgot to remove `only`

WormJim reviewed (changes requested):
Is extent a string or an array of strings?  It seems there is a mix here. Your tests are using, from the mockBib, an [array of strings](https://github.com/NYPL/discovery-front-end/pull/1891/files#diff-a0e8cba60ce3a8ebdfcdfda3244a76ad53fe222a5e99fad361f2f8605a041d00R13), and then switches to [a string](https://github.com/NYPL/discovery-front-end/pull/1891/files#diff-a0e8cba60ce3a8ebdfcdfda3244a76ad53fe222a5e99fad361f2f8605a041d00R69) for a different test block.

Given this scenario:
`
    const bib = {
      extent: ['700 sheets of woven gold;'],
    };
`  
I'd expect the returned value from `appendDimensionsToExtent` to return `bib.extent` which is an array.

Since this is the case with the above scenario it would set to the first index of `bib.extent`, [here](https://github.com/NYPL/discovery-front-end/pull/1891/files#diff-d6d9946fcb66a10d0a71c534dcb8caeeaa342a22c3e7788e16163e5af0c19076R194), to `['700 sheets of woven gold;']` effectively making `bib.extent` a 2D list. Is this your intention?

WormJim commented on test/unit/Bib.test.js:
```diff
@@ -54,6 +56,21 @@ describe('Bib', () => {
+      expect(newExtent).to.equal('99 bottles of beer; 99 x 99 cm')
+    })
+    it('should not add semicolon if it already is in extent', () => {
+      const [newExtent] = Bib.appendDimensionsToExtent({extent: '700 sheets of woven gold; ', dimensions: '1 x 1 in.'})
```
Are both `extent` and `dimensions` a single string not in a list? The `mockBib` being used above has both of these as an [array of items](https://github.com/NYPL/discovery-front-end/pull/1891/files#diff-a0e8cba60ce3a8ebdfcdfda3244a76ad53fe222a5e99fad361f2f8605a041d00R13-R14).

WormJim commented on src/server/ApiRoutes/Bib.js:
```diff
@@ -47,6 +47,23 @@ export const addHoldingDefinition = (holding) => {
+const appendDimensionsToExtent = (bib) => {
+  if (bib.extent) {
+    // Check if extent was cataloged with a semicolon already at the end
+    const semicolon = bib.extent.slice(-2) === '; ' || bib.extent.slice(-1) === ';'
```
When using the [mockBib](https://github.com/NYPL/discovery-front-end/pull/1891/files#diff-a0e8cba60ce3a8ebdfcdfda3244a76ad53fe222a5e99fad361f2f8605a041d00R13-R14) `extent` is an array or items. `Semicolon` will always be false when extent is an array since [slicing an array returns an array](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/slice).

WormJim commented on src/server/ApiRoutes/Bib.js:
```diff
@@ -47,6 +47,23 @@ export const addHoldingDefinition = (holding) => {
     .filter((data) => data.definition);
 };
 
+const appendDimensionsToExtent = (bib) => {
```
What is the expected type of `bib.extent`?

WormJim reviewed (changes requested):
I do not know if thie appendDimensionsToExtent function expects the argument of bib to have a property of extent as type string or as type array. Can this be clarified?

charmingduchess commented:
I changed all the things to be string elements of arrays, which is how all records i've seen come back from the discovery api. Also added some logic so theres a space added when necessary in between description and dimensions.

WormJim reviewed (changes requested):
Given your tests, they all pass as expected. Good Job!

I'm proposing three additional tests for you to check for.
1. 'should return undefined if no bib'
2. 'should return undefined if no extent array'
3. 'should return empty array if no extent array item' ← Really, should return extent array

Additionally, I'll also ask, what's the expected format of extent if no dimension and:
1. extent = '700 sheets of woven gold;'
2. extent = '700 sheets of woven gold; '
4. extent = '700 sheets of woven gold ;'
5. extent = '700 sheets of woven gold ; '

I'm curious if there is a necessity to remove the semicolon.

WormJim commented on src/server/ApiRoutes/Bib.js:
```diff
@@ -48,6 +48,7 @@ export const addHoldingDefinition = (holding) => {
 };
 
 const appendDimensionsToExtent = (bib) => {
+  if (!bib.extent || bib.extent.length === 0) return 
```
Can you add a test for these?

WormJim reviewed (approved):
Everything Looks good! Have a nice day! 👍🏼
